src/ast_util.py

```diff
@@ -166,6 +166,126 @@ def generic_visit(self, node: ca.Node) -> None:
     return Indices(starts, ends)
 
 
+def get_noncolliding_fn_name(ast: ca.FileAST, name: str) -> str:
+    used_names: Set[str] = set()
+    for item in ast.ext:
+        if isinstance(item, ca.Decl) and item.name:
+            used_names.add(item.name)
+        if isinstance(item, ca.FuncDef) and item.decl.name:
+            used_names.add(item.decl.name)
+    return get_noncolliding_name(used_names, name)
+
+
+def get_noncolliding_struct_name(typemap: TypeMap, name: str) -> str:
+    used_names: Set[str] = set()
+    for struct_name in typemap.struct_defs.keys():
+        used_names.add(struct_name)
+    for type_name in typemap.typedefs.keys():
+        used_names.add(type_name)
+    return get_noncolliding_name(used_names, name)
+
+
+def get_noncolliding_name(used_names: Set[str], name: str) -> str:
+    new_name = name
+    counter = 1
+    while new_name in used_names:
+        counter += 1
+        new_name = f"{new_name}{counter}"
+    return new_name
+
+
+def ast_ext_struct_idx(ast: ca.FileAST, struct: ca.Struct) -> int:
+    """Returns the index of the node in `ast.ext` that contains the definition
+    of `struct`.
+    """
+    found = False
+    class StructVisitor(ca.NodeVisitor):
+        def visit_Struct(self, node: ca.Struct) -> None:
+            nonlocal found
+            if node is struct:
+                found = True
+            self.generic_visit(node)
```

Comment:
could make this into a general "contains node" method

Reply:
Good idea! Added